yaqd_andor/_andorsdk2_ixon.py

```diff
@@ -0,0 +1,309 @@
+
+            # apply shape, mapping
+            self._channel_shapes = {"image": (int(height), int(width))}
+            self.x_ai = np.arange(int(left), int(left + width * binning), int(binning), dtype=int)[
+                None, :
+            ]
+            self.y_ai = np.arange(int(top), int(top + height * binning), int(binning), dtype=int)[
+                :, None
+            ]
+            self.binning = int(binning)
+            return
+
+    def _set_temperature(self):
+        self.sensor_cooling = self._config["sensor_cooling"]
+        if self.sensor_cooling:
+            self.sensor_temp_control = int(self._config["sensor_temperature"])
+            code = self.sdk.CoolerON()
+            if code != 20002:
+                raise ValueError(str(self.errorlookup(code)))
+            code = self.sdk.SetTemperature(self.sensor_temp_control)
+            if code != 20002:
+                raise ValueError(str(self.errorlookup(code)))
+            self.logger.info(f"Sensor is cooling.  Target temp is {self.sensor_temp_control} C.")
+            self._loop.run_in_executor(None, self._check_temp_stabilized)
+        else:
+            code, self.sensor_temp = self.sdk.GetTemperature()
+            self.logger.info(f"Sensor is not cooled.  Current temp is {self.sensor_temp} C.")
+
+    def _check_temp_stabilized(self):
+        code, self.sensor_temp = self.sdk.GetTemperature()
+        diff = float(self.sensor_temp_control) - float(self.sensor_temp)
+        while (
+            np.abs(diff) > 1.0 and self.stop_update is False
+        ):  # this is a tolerance, and is subject to change
+            self.logger.info(
+                f"Sensor is cooling.  Target: {self.sensor_temp_control} C.  Current: {self.sensor_temp:0.2f} C."
+            )
+            sleep(3)
+            code, self.sensor_temp = self.sdk.GetTemperature()
+            diff = float(self.sensor_temp) - float(self.sensor_temp_control)
+        if np.abs(diff) <= 1.0:
+            self.logger.info("Sensor temp is stabilized.")
+
+    def get_dependent_hardware(self):
+        dependents = dict()
+        if self.spec_client is not None:
+            dependents = {"monochromator": self._spec_position}
+        return dependents
+
+    def set_exposure_time(self, exposure_time):
+        # Sets the exposure time in seconds (float)
+        self.stop_update = True
+        while self._busy == True:
+            sleep(0.10)
+        code = self.sdk.SetExposureTime(float(exposure_time))
+        if code != 20002:
+            raise ValueError(str(self.errorlookup(code)))
+        else:
+            self.exposure_time = exposure_time
+            self.logger.info(f"New exposure time is {self.exposure_time} sec.")
+        self.stop_update = False
+        return code
+
+    def get_exposure_time(self):
+        return self.exposure_time
+
+    def close(self):
+        self.stop_update = True
+        while self._busy == True:
+            sleep(0.50)
```

Comment:
`close` is a synchronous function, so once you start running it, you block everything else from running.  This while loop will never exit if busy is true.  Remove.

Reply:
Isn't `update_state` still running on an asynchronous thread irrespective of close being called?   That thread is modifying _busy

Reply:
`update_state` is running on another thread, but this code block never frees up the processor for that thread to run.  You would need asyncio.sleep to allow another thread such as update_state to execute.

Reply:
I guess I am missing something again regarding sync vs async loops.  I would have thought the async threads are assigned time by the GIL automatically, that things might occur in the sync loop that could lock up the GIL and stop  it from going through the threads in turn.

Reply:
correct--sleep locks the thread. async-style statements are the only way to free up threads (usually in the form of an `await` statement).

Reply:
ok